apiserver/facades/agent/provisioner/provisioner.go

```diff
@@ -1424,16 +1427,23 @@ func (api *ProvisionerAPI) SetCharmProfiles(ctx context.Context, args params.Set
 		return params.ErrorResults{}, errors.Trace(err)
 	}
 	for i, a := range args.Args {
-		results[i].Error = apiservererrors.ServerError(api.setOneMachineCharmProfiles(a.Entity.Tag, a.Profiles, canAccess))
+		results[i].Error = apiservererrors.ServerError(api.setOneMachineCharmProfiles(ctx, a.Entity.Tag, a.Profiles, canAccess))
 	}
 	return params.ErrorResults{Results: results}, nil
 }
 
-func (api *ProvisionerAPI) setOneMachineCharmProfiles(machineTag string, profiles []string, canAccess common.AuthFunc) error {
+func (api *ProvisionerAPI) setOneMachineCharmProfiles(ctx context.Context, machineTag string, profiles []string, canAccess common.AuthFunc) error {
 	mTag, err := names.ParseMachineTag(machineTag)
 	if err != nil {
 		return errors.Trace(err)
 	}
+	machineUUID, err := api.machineService.GetMachineUUID(ctx, coremachine.Name(mTag.Id()))
```

Comment:
question: where access to the machine checked here, like for getMachine below?

Reply:
So, since we are still double writing there is no need to rewrite the access check (it's already checked every time we retrieve the machine with `api.getMachine()`. 
It's a good remark though and we should be diligent when we fully remove machines to make sure we keep these checks on each api endpoint.